sourcetool/pkg/auth/implementation.go

```diff
@@ -0,0 +1,191 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+)
+
+//counterfeiter:generate . authenticatorImplementation
+type authenticatorImplementation interface {
+	requestDeviceCode(context.Context) (*DeviceCodeResponse, error)
+	openBrowser(string) error
+	pollForToken(context.Context, string, time.Duration) (string, error)
+	checkTokenStatus(context.Context, string) (string, error)
+	persistToken(token string) error
+	readToken() (string, error)
+}
+
+type defaultImplementation struct{}
+
+// pollForToken polls GitHub periodically until the device flow is done and
+// the token is issued.
+func (di *defaultImplementation) pollForToken(ctx context.Context, deviceCode string, interval time.Duration) (string, error) {
+	for {
+		token, err := di.checkTokenStatus(ctx, deviceCode)
+		if err != nil {
+			if strings.Contains(err.Error(), "authorization_pending") {
+				time.Sleep(interval)
+				continue
+			}
+			if strings.Contains(err.Error(), "slow_down") {
+				interval += 5 * time.Second
+				time.Sleep(interval)
+				continue
+			}
+			return "", err
+		}
+		return token, nil
+	}
+}
+
+// checkTokenStatus checks if the device flow authorization is done and returns
+// the token if ready.
+func (di *defaultImplementation) checkTokenStatus(ctx context.Context, deviceCode string) (string, error) {
+	data := url.Values{}
+	data.Set("client_id", oauthClientID)
+	data.Set("device_code", deviceCode)
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, tokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return "", fmt.Errorf("failed to create token request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Set("Accept", "application/json")
+
+	httpClient := &http.Client{Timeout: 30 * time.Second}
```

Comment:
Is there an advantage to setting a client timeout vs. setting the timeout via the request's Context? If you dont' set the timeout here but rely on the Context, the caller has more control over the timeout.

Reply:
It might be good to still have more granular timeouts so a single request doesn't eat up the whole 300s timeout. I'll defer to your judgement though.